Beans not injected into a service impl with `@CXFEndpoint("/my-path")`

Today I did the update to quarkus-cxf 3.11.0 and every thing runs fine. Then I removed `quarkus.cxf.endpoint."/my-path".implementor` from the properties file and added the new annotation @CXFEndpoint("/my-path") to my service. After that all unit tests was failing because the injected beans in my service was null.

For the record, the stack trace is

```
WARN  [org.apa.cxf.pha.PhaseInterceptorChain] (executor-thread-1) Application {https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/test}HelloService#{https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/test}hello has thrown exception, unwinding now: org.apache.cxf.interceptor.Fault: Cannot invoke "io.quarkiverse.cxf.it.annotation.cxfendpoint.HelloBean.hello(String)" because "this.helloBean" is null
[INFO] [stdout]         at org.apache.cxf.service.invoker.AbstractInvoker.createFault(AbstractInvoker.java:162)
[INFO] [stdout]         at org.apache.cxf.jaxws.AbstractJAXWSMethodInvoker.createFault(AbstractJAXWSMethodInvoker.java:266)
[INFO] [stdout]         at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:128)
[INFO] [stdout]         at org.apache.cxf.jaxws.AbstractJAXWSMethodInvoker.invoke(AbstractJAXWSMethodInvoker.java:231)
[INFO] [stdout]         at org.apache.cxf.jaxws.JAXWSMethodInvoker.invoke(JAXWSMethodInvoker.java:84)
[INFO] [stdout]         at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:74)
[INFO] [stdout]         at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
[INFO] [stdout]         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
[INFO] [stdout]         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[INFO] [stdout]         at org.apache.cxf.interceptor.ServiceInvokerInterceptor$2.run(ServiceInvokerInterceptor.java:126)
[INFO] [stdout]         at org.apache.cxf.workqueue.SynchronousExecutor.execute(SynchronousExecutor.java:37)
[INFO] [stdout]         at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:131)
[INFO] [stdout]         at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
[INFO] [stdout]         at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
[INFO] [stdout]         at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:265)
[INFO] [stdout]         at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:233)
[INFO] [stdout]         at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:207)
[INFO] [stdout]         at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:159)
[INFO] [stdout]         at io.quarkiverse.cxf.transport.CxfHandler.process(CxfHandler.java:248)
[INFO] [stdout]         at io.quarkiverse.cxf.transport.CxfHandler.handle(CxfHandler.java:185)
[INFO] [stdout]         at io.quarkiverse.cxf.transport.CxfHandler.handle(CxfHandler.java:50)
[INFO] [stdout]         at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
[INFO] [stdout]         at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:191)
[INFO] [stdout]         at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:279)
[INFO] [stdout]         at io.vertx.core.impl.ContextImpl.lambda$internalExecuteBlocking$2(ContextImpl.java:210)
[INFO] [stdout]         at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:599)
[INFO] [stdout]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[INFO] [stdout]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[INFO] [stdout]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[INFO] [stdout]         at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
[INFO] [stdout]         at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
[INFO] [stdout]         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
[INFO] [stdout]         at java.base/java.lang.Thread.run(Thread.java:1583)
[INFO] [stdout] Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkiverse.cxf.it.annotation.cxfendpoint.HelloBean.hello(String)" because "this.helloBean" is null
[INFO] [stdout]         at io.quarkiverse.cxf.it.annotation.cxfendpoint.PathAnnotationHelloServiceWithBean.hello(PathAnnotationHelloServiceWithBean.java:20)
[INFO] [stdout]         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[INFO] [stdout]         at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[INFO] [stdout]         at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)
[INFO] [stdout]         at org.apache.cxf.jaxws.JAXWSMethodInvoker.performInvocation(JAXWSMethodInvoker.java:65)
[INFO] [stdout]         at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
[INFO] [stdout]         ... 30 more
```

--- a/docs/modules/ROOT/examples/soap-service/HelloServiceImpl.java
+++ b/docs/modules/ROOT/examples/soap-service/HelloServiceImpl.java
@@ -6,7 +6,7 @@ import jakarta.jws.WebService;
 /**
  * The simplest Hello service implementation.
  */
-@WebService(serviceName = "HelloService", portName = "HelloServicePort")
+@WebService(serviceName = "HelloService")
 public class HelloServiceImpl implements HelloService {
 
     @WebMethod

--- a/docs/modules/ROOT/examples/soap-service/application.properties
+++ b/docs/modules/ROOT/examples/soap-service/application.properties
@@ -63,7 +63,7 @@ quarkus.cxf.client."greetingclient".service-interface = io.quarkiverse.cxf.it.se
 
 # java2ws
 # tag::generate-wsdl-from-java.adoc[]
-quarkus.cxf.java2ws.includes = io.quarkiverse.cxf.it.server.HelloService,io.quarkiverse.cxf.it.server.FaultyHelloService
+quarkus.cxf.java2ws.includes = io.quarkiverse.cxf.it.server.HelloServiceImpl,io.quarkiverse.cxf.it.server.FaultyHelloServiceImpl
 quarkus.cxf.java2ws.wsdl-name-template = %TARGET_DIR%/Java2wsTest/%SIMPLE_CLASS_NAME%-from-java2ws.wsdl
 # end::generate-wsdl-from-java.adoc[]
 

--- a/docs/modules/ROOT/pages/user-guide/generate-wsdl-from-java.adoc
+++ b/docs/modules/ROOT/pages/user-guide/generate-wsdl-from-java.adoc
@@ -25,17 +25,17 @@ in the source tree of {quarkus-cxf-project-name}
 include::example$soap-service/application.properties[tag=generate-wsdl-from-java.adoc]
 ----
 
-Here we have instructed `java2ws` to generate WSDLs for two interfaces,
-namely `HelloService` and `FaultyHelloService`.
+Here we have instructed `java2ws` to generate WSDLs for two service classes,
+namely `HelloServiceImpl` and `FaultyHelloServiceImpl`.
 
 [NOTE]
 .Annotations
 ====
-Note that the Service interfaces must be annotated with `{javaxOrJakartaPackagePrefix}.xml.ws.WebService` to be selected for `java2ws` processing.
+Note that the service classes must be annotated with `{javaxOrJakartaPackagePrefix}.xml.ws.WebService` to be selected for `java2ws` processing.
 ====
 
-The two generated WSDL documents will be stored as `target/Java2wsTest/FaultyHelloService-from-java2ws.wsdl`
-and `target/Java2wsTest/HelloService-from-java2ws.wsdl` respectively.
+The two generated WSDL documents will be stored as `target/Java2wsTest/FaultyHelloServiceImpl-from-java2ws.wsdl`
+and `target/Java2wsTest/HelloServiceImpl-from-java2ws.wsdl` respectively.
 
 NOTE: Unlike `wsdl2java` which is executed within Quarkus source generation phase,
       `java2ws` is a part Quarkus augmentation that happens after compilation.

--- a/extensions/core/deployment/src/main/java/io/quarkiverse/cxf/deployment/CxfDeploymentUtils.java
+++ b/extensions/core/deployment/src/main/java/io/quarkiverse/cxf/deployment/CxfDeploymentUtils.java
@@ -1,6 +1,7 @@
 package io.quarkiverse.cxf.deployment;
 
 import java.util.Collection;
+import java.util.List;
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 
@@ -11,7 +12,9 @@ import org.apache.cxf.jaxws.JaxWsClientFactoryBean;
 import org.apache.cxf.jaxws.JaxWsServerFactoryBean;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
+import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
 
 public class CxfDeploymentUtils {
@@ -19,7 +22,7 @@ public class CxfDeploymentUtils {
     private CxfDeploymentUtils() {
     }
 
-    public static Stream<AnnotationInstance> cxfEndpointAnnotations(IndexView index) {
+    public static Stream<AnnotationInstance> methodsWithCxfEndpointAnnotations(IndexView index) {
         return Stream.of(CxfDotNames.CXF_ENDPOINT_ANNOTATION)
                 .map(index::getAnnotations)
                 .flatMap(Collection::stream)
@@ -31,6 +34,35 @@ public class CxfDeploymentUtils {
                 .map(index::getAnnotations)
                 .flatMap(Collection::stream)
                 .filter(annotation -> annotation.target().kind() == AnnotationTarget.Kind.CLASS);
+    }
+
+    public static String findSei(IndexView index, ClassInfo wsClassInfo) {
+
+        {
+            final AnnotationInstance webServiceAnnotation = wsClassInfo.annotation(CxfDotNames.WEBSERVICE_ANNOTATION);
+            if (webServiceAnnotation != null) {
+                if (wsClassInfo.isInterface()) {
+                    return wsClassInfo.name().toString();
+                }
+                final AnnotationValue val = webServiceAnnotation.value("endpointInterface");
+                if (val != null) {
+                    final String result = val.asString();
+                    if (!result.isEmpty()) {
+                        return result;
+                    }
+                }
+            }
+        }
+
+        final List<DotName> interfaces = wsClassInfo.interfaceNames();
+        for (DotName interfaceName : interfaces) {
+            final ClassInfo interfaceInfo = index.getClassByName(interfaceName);
+            final AnnotationInstance webServiceAnnotation = interfaceInfo.annotation(CxfDotNames.WEBSERVICE_ANNOTATION);
+            if (webServiceAnnotation != null) {
+                return interfaceInfo.name().toString();
+            }
+        }
+        return wsClassInfo.name().toString();
     }
 
     /**

--- a/extensions/core/deployment/src/main/java/io/quarkiverse/cxf/deployment/CxfEndpointImplementationBuildItem.java
+++ b/extensions/core/deployment/src/main/java/io/quarkiverse/cxf/deployment/CxfEndpointImplementationBuildItem.java
@@ -9,18 +9,31 @@ import io.quarkiverse.cxf.CXFRecorder.BeanLookupStrategy;
  */
 public final class CxfEndpointImplementationBuildItem extends AbstractEndpointBuildItem {
 
+    private final String sei;
     private final String implementor;
     private final boolean provider;
     private final String relativePath;
     private final BeanLookupStrategy beanLookupStrategy;
 
-    public CxfEndpointImplementationBuildItem(String implementor, String soapBinding, String wsNamespace,
-            String wsName, boolean provider, String relativePath, BeanLookupStrategy beanLookupStrategy) {
+    public CxfEndpointImplementationBuildItem(
+            String sei,
+            String implementor,
+            String soapBinding,
+            String wsNamespace,
+            String wsName,
+            boolean provider,
+            String relativePath,
+            BeanLookupStrategy beanLookupStrategy) {
         super(soapBinding, wsNamespace, wsName);
+        this.sei = sei;
         this.implementor = Objects.requireNonNull(implementor, "implementor cannot be null");
         this.provider = provider;
         this.relativePath = relativePath;
         this.beanLookupStrategy = beanLookupStrategy;
+    }
+
+    public String getSei() {
+        return sei;
     }
 
     public String getImplementor() {

--- a/extensions/core/deployment/src/main/java/io/quarkiverse/cxf/deployment/CxfEndpointImplementationProcessor.java
+++ b/extensions/core/deployment/src/main/java/io/quarkiverse/cxf/deployment/CxfEndpointImplementationProcessor.java
@@ -81,7 +81,15 @@ public class CxfEndpointImplementationProcessor {
 
                         final AnnotationInstance cxfEndpointAnnotation = wsClassInfo
                                 .declaredAnnotation(CxfDotNames.CXF_ENDPOINT_ANNOTATION);
-                        final String relPath = cxfEndpointAnnotation != null ? cxfEndpointAnnotation.value().asString() : null;
+                        final String relPath;
+                        final BeanLookupStrategy beanLookupStrategy;
+                        if (cxfEndpointAnnotation != null) {
+                            relPath = cxfEndpointAnnotation.value().asString();
+                            beanLookupStrategy = BeanLookupStrategy.TYPE_WITH_CXFENDPOINT_ANNOTATION;
+                        } else {
+                            relPath = null;
+                            beanLookupStrategy = BeanLookupStrategy.TYPE;
+                        }
 
                         submitImpl(
                                 endpointImplementations,
@@ -90,10 +98,11 @@ public class CxfEndpointImplementationProcessor {
                                 reflectives,
                                 annotation,
                                 wsClassInfo,
+                                CxfDeploymentUtils.findSei(index, wsClassInfo),
                                 impl,
                                 hasWebServiceProviderAnnotation,
                                 relPath,
-                                BeanLookupStrategy.TYPE);
+                                beanLookupStrategy);
 
                     } else if (Modifier.isInterface(wsClassInfo.flags())) {
                         String cl = wsClassInfo.name().toString();
@@ -107,12 +116,13 @@ public class CxfEndpointImplementationProcessor {
                     }
 
                 });
-        CxfDeploymentUtils.cxfEndpointAnnotations(index)
+        CxfDeploymentUtils.methodsWithCxfEndpointAnnotations(index)
                 .forEach(annotation -> {
                     final MethodInfo methodInfo = annotation.target().asMethod();
 
-                    final String impl = methodInfo.returnType().name().toString();
-                    final ClassInfo wsClassInfo = index.getClassByName(methodInfo.returnType().name());
+                    final DotName returnType = methodInfo.returnType().name();
+                    final String impl = returnType.toString();
+                    final ClassInfo wsClassInfo = index.getClassByName(returnType);
                     AnnotationInstance wsAnnot = wsClassInfo.declaredAnnotation(CxfDotNames.WEBSERVICE_ANNOTATION);
                     final boolean hasWebServiceProviderAnnotation = wsAnnot != null;
                     if (wsAnnot == null) {
@@ -131,10 +141,11 @@ public class CxfEndpointImplementationProcessor {
                             reflectives,
                             wsAnnot,
                             wsClassInfo,
+                            CxfDeploymentUtils.findSei(index, wsClassInfo),
                             impl,
                             hasWebServiceProviderAnnotation,
                             annotation.value().asString(),
-                            BeanLookupStrategy.TYPE_WITH_CXFENDPOINT_ANNOTATION);
+                            BeanLookupStrategy.METHOD_WITH_CXFENDPOINT_ANNOTATION);
                 });
 
         reflectiveClass
@@ -146,6 +157,7 @@ public class CxfEndpointImplementationProcessor {
             BuildProducer<ServiceSeiBuildItem> serviceSeis,
             BuildProducer<AdditionalBeanBuildItem> additionalBeans, Set<String> reflectives, AnnotationInstance annotation,
             ClassInfo wsClassInfo,
+            String sei,
             String impl,
             boolean hasWebServiceProviderAnnotation,
             String relativePathFromCxfEndpointAnnotation,
@@ -166,7 +178,7 @@ public class CxfEndpointImplementationProcessor {
 
         final String wsName = Optional.ofNullable(annotation.value("serviceName"))
                 .map(AnnotationValue::asString)
-                .orElse(impl.contains(".") ? impl.substring(impl.lastIndexOf('.') + 1) : impl);
+                .orElse(sei.contains(".") ? sei.substring(sei.lastIndexOf('.') + 1) : sei);
 
         String soapBinding = Optional
                 .ofNullable(wsClassInfo.declaredAnnotation(CxfDotNames.BINDING_TYPE_ANNOTATION))
@@ -175,6 +187,7 @@ public class CxfEndpointImplementationProcessor {
 
         endpointImplementations.produce(
                 new CxfEndpointImplementationBuildItem(
+                        sei,
                         impl,
                         soapBinding,
                         wsNamespace,
@@ -204,21 +217,27 @@ public class CxfEndpointImplementationProcessor {
                 .collect(Collectors.toList());
         if (!cxfEndpoints.isEmpty()) {
             RuntimeValue<Map<String, List<ServletConfig>>> implementorToCfgMap = recorder.implementorToCfgMap(cxfConfig);
+
+            final ClassLoader cl = Thread.currentThread().getContextClassLoader();
             for (CxfEndpointImplementationBuildItem cxfWebService : cxfEndpoints) {
-                recorder.addCxfServletInfo(
-                        infos,
-                        implementorToCfgMap,
-                        fixedConfig.path(),
-                        cxfWebService.getImplementor(),
-                        cxfConfig,
-                        cxfWebService.getWsName(),
-                        cxfWebService.getWsNamespace(),
-                        cxfWebService.getSoapBinding(),
-                        cxfWebService.getImplementor(),
-                        cxfWebService.isProvider(),
-                        cxfWebService.getRelativePath(),
-                        cxfWebService.getBeanLookupStrategy());
-                requestors.add(cxfWebService.getImplementor());
+                try {
+                    recorder.addCxfServletInfo(
+                            infos,
+                            implementorToCfgMap,
+                            fixedConfig.path(),
+                            cl.loadClass(cxfWebService.getSei()),
+                            cxfConfig,
+                            cxfWebService.getWsName(),
+                            cxfWebService.getWsNamespace(),
+                            cxfWebService.getSoapBinding(),
+                            cl.loadClass(cxfWebService.getImplementor()),
+                            cxfWebService.isProvider(),
+                            cxfWebService.getRelativePath(),
+                            cxfWebService.getBeanLookupStrategy());
+                    requestors.add(cxfWebService.getImplementor());
+                } catch (ClassNotFoundException e) {
+                    throw new RuntimeException(e);
+                }
             }
         }
         if (!requestors.isEmpty()) {

--- a/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/CXFRecorder.java
+++ b/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/CXFRecorder.java
@@ -48,46 +48,52 @@ public class CXFRecorder {
     public enum BeanLookupStrategy {
         TYPE {
             @Override
-            public Supplier<Object> createLookUp(String type, String path) {
-                return () -> CXFRuntimeUtils.getInstance(type, false);
+            public Supplier<Object> createLookUp(Class<?> sei, Class<?> wsImplementor, String path) {
+                return () -> CXFRuntimeUtils.getInstance(wsImplementor);
             }
+        },
+        METHOD_WITH_CXFENDPOINT_ANNOTATION {
+            @Override
+            public Supplier<Object> createLookUp(Class<?> sei, Class<?> wsImplementor, String path) {
+                return createLookupInternal(sei, path);
+            }
+
         },
         TYPE_WITH_CXFENDPOINT_ANNOTATION {
             @Override
-            public Supplier<Object> createLookUp(String type, String path) {
-                return () -> {
-                    try {
-                        Class<?> wsImplementorClass = Class.forName(type, false,
-                                Thread.currentThread().getContextClassLoader());
-                        Object result = Arc.container()
-                                .instance(wsImplementorClass,
-                                        new CXFEndpoint.CXFEndpointLiteral(path))
-                                .get();
-                        if (result == null) {
-                            throw new IllegalStateException("Could get bean of type " + type + " qualified by @"
-                                    + CXFEndpoint.class.getName() + "(\"" + path + "\")");
-                        }
-                        return result;
-                    } catch (ClassNotFoundException e) {
-                        throw new RuntimeException("Could not load class " + type, e);
-                    }
-                };
+            public Supplier<Object> createLookUp(Class<?> sei, Class<?> wsImplementor, String path) {
+                return createLookupInternal(wsImplementor, path);
             }
+
         };
 
-        public abstract Supplier<Object> createLookUp(String type, String path);
+        private static Supplier<Object> createLookupInternal(Class<?> type, String path) {
+            return () -> {
+                Object result = Arc.container()
+                        .instance(type,
+                                new CXFEndpoint.CXFEndpointLiteral(path))
+                        .get();
+                if (result == null) {
+                    throw new IllegalStateException("Could get bean of type " + type + " qualified by @"
+                            + CXFEndpoint.class.getName() + "(\"" + path + "\")");
+                }
+                return result;
+            };
+        }
+
+        public abstract Supplier<Object> createLookUp(Class<?> sei, Class<?> wsImplementor, String path);
     }
 
     public void addCxfServletInfo(
             RuntimeValue<CXFServletInfos> runtimeInfos,
             RuntimeValue<Map<String, List<ServletConfig>>> implementorToCfg,
             String path,
-            String sei,
+            Class<?> sei,
             CxfConfig cxfConfig,
             String serviceName,
             String serviceTargetNamepsace,
             String soapBinding,
-            String wsImplementor,
+            Class<?> wsImplementor,
             Boolean isProvider,
             String relativePathFromCxfEndpointAnnotation,
             BeanLookupStrategy beanLookupStrategy) {
@@ -95,6 +101,7 @@ public class CXFRecorder {
         CXFServletInfos infos = runtimeInfos.getValue();
 
         switch (beanLookupStrategy) {
+            case METHOD_WITH_CXFENDPOINT_ANNOTATION:
             case TYPE_WITH_CXFENDPOINT_ANNOTATION: {
                 final CxfEndpointConfig cxfEndPointConfig = cxfConfig.endpoints().get(relativePathFromCxfEndpointAnnotation);
                 final CXFServletInfo info = createServletInfo(
@@ -107,7 +114,7 @@ public class CXFRecorder {
                         cxfEndPointConfig,
                         relativePathFromCxfEndpointAnnotation,
                         isProvider,
-                        beanLookupStrategy.createLookUp(sei, relativePathFromCxfEndpointAnnotation));
+                        beanLookupStrategy.createLookUp(sei, wsImplementor, relativePathFromCxfEndpointAnnotation));
                 infos.add(info);
                 return;
             }
@@ -125,11 +132,11 @@ public class CXFRecorder {
                             cxfEndPointConfig,
                             relativePathFromCxfEndpointAnnotation,
                             isProvider,
-                            beanLookupStrategy.createLookUp(wsImplementor, relativePathFromCxfEndpointAnnotation));
+                            beanLookupStrategy.createLookUp(sei, wsImplementor, relativePathFromCxfEndpointAnnotation));
                     infos.add(info);
                     return;
                 }
-                List<ServletConfig> cfgs = implementorToCfg.getValue().get(wsImplementor);
+                List<ServletConfig> cfgs = implementorToCfg.getValue().get(wsImplementor.getName());
                 if (cfgs != null) {
                     for (ServletConfig cfg : cfgs) {
                         CxfEndpointConfig cxfEndPointConfig = cfg.config;
@@ -144,12 +151,12 @@ public class CXFRecorder {
                                 cxfEndPointConfig,
                                 relativePath,
                                 isProvider,
-                                beanLookupStrategy.createLookUp(wsImplementor, relativePathFromCxfEndpointAnnotation));
+                                beanLookupStrategy.createLookUp(sei, wsImplementor, relativePathFromCxfEndpointAnnotation));
                         infos.add(info);
                     }
                 } else {
                     if (serviceName == null || serviceName.isEmpty()) {
-                        serviceName = sei.toLowerCase();
+                        serviceName = sei.getName().toLowerCase();
                         if (serviceName.contains(".")) {
                             serviceName = serviceName.substring(serviceName.lastIndexOf('.') + 1);
                         }
@@ -165,7 +172,7 @@ public class CXFRecorder {
                             null,
                             relativePath,
                             isProvider,
-                            beanLookupStrategy.createLookUp(wsImplementor, relativePathFromCxfEndpointAnnotation));
+                            beanLookupStrategy.createLookUp(sei, wsImplementor, relativePathFromCxfEndpointAnnotation));
                     infos.add(info);
                 }
                 return;
@@ -198,11 +205,11 @@ public class CXFRecorder {
 
     private static CXFServletInfo createServletInfo(
             String path,
-            String sei,
+            Class<?> sei,
             String serviceName,
             String serviceTargetNamespace,
             String soapBinding,
-            String wsImplementor,
+            Class<?> implementor,
             CxfEndpointConfig cxfEndPointConfig,
             String relativePath,
             Boolean isProvider,
@@ -210,7 +217,7 @@ public class CXFRecorder {
         CXFServletInfo cfg = new CXFServletInfo(
                 path,
                 relativePath,
-                wsImplementor,
+                implementor,
                 sei,
                 cxfEndPointConfig != null ? cxfEndPointConfig.wsdlPath().orElse(null) : null,
                 serviceName,

--- a/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/CXFRuntimeUtils.java
+++ b/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/CXFRuntimeUtils.java
@@ -8,7 +8,6 @@ import jakarta.enterprise.inject.UnsatisfiedResolutionException;
 import jakarta.enterprise.inject.spi.CDI;
 
 import io.quarkus.arc.Arc;
-import io.quarkus.arc.Subclass;
 
 public class CXFRuntimeUtils {
 
@@ -111,18 +110,6 @@ public class CXFRuntimeUtils {
                 destination.add(item);
             }
         }
-    }
-
-    /**
-     * @param beanType a bean type, or a subclass thereof wrapped by the CDI container
-     * @return the proper bean type, unwrapped if needed
-     */
-    public static Class<?> unwrapBeanType(Class<?> beanType) {
-        Objects.requireNonNull(beanType, "bean type");
-        if (Subclass.class.isAssignableFrom(beanType)) {
-            return beanType.getSuperclass();
-        }
-        return beanType;
     }
 
     private static Class<?> loadClass(String className) {

--- a/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/CXFServletInfo.java
+++ b/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/CXFServletInfo.java
@@ -12,14 +12,14 @@ import io.netty.util.internal.shaded.org.jctools.queues.MessagePassingQueue.Supp
 public class CXFServletInfo {
     private final String relativePath;
     private final String path;
-    private final String className;
+    private final Class<?> implementor;
     private final List<String> inInterceptors;
     private final List<String> outInterceptors;
     private final List<String> outFaultInterceptors;
     private final List<String> inFaultInterceptors;
     private final List<String> features;
     private final List<String> handlers;
-    private final String sei;
+    private final Class<?> sei;
     private final String wsdlPath;
     private final String serviceName;
     private final String serviceTargetNamespace;
@@ -34,8 +34,8 @@ public class CXFServletInfo {
     public CXFServletInfo(
             String path,
             String relativePath,
-            String className,
-            String sei,
+            Class<?> implementor,
+            Class<?> sei,
             String wsdlPath,
             String serviceName,
             String serviceTargetNamespace,
@@ -47,7 +47,7 @@ public class CXFServletInfo {
         LOGGER.trace("new CXFServletInfo");
         this.path = path;
         this.relativePath = relativePath;
-        this.className = className;
+        this.implementor = implementor;
         this.inInterceptors = new ArrayList<>();
         this.outInterceptors = new ArrayList<>();
         this.outFaultInterceptors = new ArrayList<>();
@@ -65,8 +65,15 @@ public class CXFServletInfo {
         this.beanLookup = beanLookup;
     }
 
+    /**
+     * @return fully qualified Service Implementation class name
+     */
     public String getClassName() {
-        return className;
+        return implementor.getName();
+    }
+
+    public Class<?> getImplementor() {
+        return implementor;
     }
 
     public String getPath() {
@@ -81,7 +88,7 @@ public class CXFServletInfo {
         return wsdlPath;
     }
 
-    public String getSei() {
+    public Class<?> getSei() {
         return sei;
     }
 
@@ -170,6 +177,6 @@ public class CXFServletInfo {
 
     @Override
     public String toString() {
-        return "Web Service " + className + " on " + path;
+        return "Web Service " + implementor.getName() + " on " + path;
     }
 }

--- a/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/transport/CxfHandler.java
+++ b/extensions/core/runtime/src/main/java/io/quarkiverse/cxf/transport/CxfHandler.java
@@ -108,7 +108,7 @@ public class CxfHandler implements Handler<RoutingContext> {
             final String endpointString = "endpoint " + servletInfo.getPath();
             final Object instanceService = servletInfo.lookupBean();
             if (instanceService != null) {
-                final Class<?> instanceType = io.quarkiverse.cxf.CXFRuntimeUtils.unwrapBeanType(instanceService.getClass());
+                final Class<?> instanceType = servletInfo.getImplementor();
                 final QuarkusRuntimeJaxWsServiceFactoryBean jaxWsServiceFactoryBean = new QuarkusRuntimeJaxWsServiceFactoryBean(
                         new JaxWsImplementorInfo(instanceType));
                 final JaxWsServerFactoryBean jaxWsServerFactoryBean = new QuarkusJaxWsServerFactoryBean(jaxWsServiceFactoryBean,

--- a/integration-tests/client-server/src/main/java/io/quarkiverse/cxf/it/annotation/cxfendpoint/HelloBean.java
+++ b/integration-tests/client-server/src/main/java/io/quarkiverse/cxf/it/annotation/cxfendpoint/HelloBean.java
@@ -1,0 +1,10 @@
+package io.quarkiverse.cxf.it.annotation.cxfendpoint;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class HelloBean {
+    public String hello(String person) {
+        return "Hello " + person + " from HelloBean!";
+    }
+}

--- a/integration-tests/client-server/src/main/java/io/quarkiverse/cxf/it/annotation/cxfendpoint/PathAnnotationHelloServiceWithBean.java
+++ b/integration-tests/client-server/src/main/java/io/quarkiverse/cxf/it/annotation/cxfendpoint/PathAnnotationHelloServiceWithBean.java
@@ -1,0 +1,20 @@
+package io.quarkiverse.cxf.it.annotation.cxfendpoint;
+
+import jakarta.inject.Inject;
+import jakarta.jws.WebService;
+
+import io.quarkiverse.cxf.annotation.CXFEndpoint;
+import io.quarkiverse.cxf.it.HelloService;
+
+@CXFEndpoint("/path-annotation-with-bean")
+@WebService(serviceName = "HelloService", targetNamespace = HelloService.NS)
+public class PathAnnotationHelloServiceWithBean implements HelloService {
+
+    @Inject
+    HelloBean helloBean;
+
+    @Override
+    public String hello(String person) {
+        return helloBean.hello(person);
+    }
+}

--- a/integration-tests/client-server/src/test/java/io/quarkiverse/cxf/it/annotation/cxfendpoint/CXFEndpointAnnotationTest.java
+++ b/integration-tests/client-server/src/test/java/io/quarkiverse/cxf/it/annotation/cxfendpoint/CXFEndpointAnnotationTest.java
@@ -19,4 +19,13 @@ public class CXFEndpointAnnotationTest {
         Assertions.assertThat(client.hello("Joe")).isEqualTo("Hello Joe from PathAnnotationHelloServiceImpl!");
     }
 
+    @Test
+    void annotatedImplementationTypeWithBean() {
+        final HelloService client = QuarkusCxfClientTestUtil.getClient(
+                "https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/test",
+                HelloService.class,
+                "/soap/path-annotation-with-bean");
+        Assertions.assertThat(client.hello("Joe")).isEqualTo("Hello Joe from HelloBean!");
+    }
+
 }

--- a/integration-tests/server/src/main/java/io/quarkiverse/cxf/it/server/FaultyHelloServiceImpl.java
+++ b/integration-tests/server/src/main/java/io/quarkiverse/cxf/it/server/FaultyHelloServiceImpl.java
@@ -6,7 +6,7 @@ import jakarta.jws.WebService;
 /**
  * A Web service always throwing an exception.
  */
-@WebService(serviceName = "FaultyHelloService", portName = "FaultyHelloServicePort")
+@WebService(serviceName = "FaultyHelloService")
 public class FaultyHelloServiceImpl implements FaultyHelloService {
 
     @WebMethod

--- a/integration-tests/server/src/main/java/io/quarkiverse/cxf/it/server/HelloServiceImpl.java
+++ b/integration-tests/server/src/main/java/io/quarkiverse/cxf/it/server/HelloServiceImpl.java
@@ -6,7 +6,7 @@ import jakarta.jws.WebService;
 /**
  * The simplest Hello service implementation.
  */
-@WebService(serviceName = "HelloService", portName = "HelloServicePort")
+@WebService(serviceName = "HelloService")
 public class HelloServiceImpl implements HelloService {
 
     @WebMethod

--- a/integration-tests/server/src/main/resources/application.properties
+++ b/integration-tests/server/src/main/resources/application.properties
@@ -63,7 +63,7 @@ quarkus.cxf.client."greetingclient".service-interface = io.quarkiverse.cxf.it.se
 
 # java2ws
 # tag::generate-wsdl-from-java.adoc[]
-quarkus.cxf.java2ws.includes = io.quarkiverse.cxf.it.server.HelloService,io.quarkiverse.cxf.it.server.FaultyHelloService
+quarkus.cxf.java2ws.includes = io.quarkiverse.cxf.it.server.HelloServiceImpl,io.quarkiverse.cxf.it.server.FaultyHelloServiceImpl
 quarkus.cxf.java2ws.wsdl-name-template = %TARGET_DIR%/Java2wsTest/%SIMPLE_CLASS_NAME%-from-java2ws.wsdl
 # end::generate-wsdl-from-java.adoc[]
 

--- a/integration-tests/server/src/test/java/io/quarkiverse/cxf/it/server/Java2wsTest.java
+++ b/integration-tests/server/src/test/java/io/quarkiverse/cxf/it/server/Java2wsTest.java
@@ -35,13 +35,13 @@ public class Java2wsTest {
 
     @Test
     public void java2WsHelloService() throws IOException, ParserConfigurationException, SAXException, TransformerException {
-        java2Ws("hello", "HelloService");
+        java2Ws("hello", "HelloServiceImpl");
     }
 
     @Test
     public void java2WsFaultyHelloService()
             throws IOException, ParserConfigurationException, SAXException, TransformerException {
-        java2Ws("faulty-hello", "FaultyHelloService");
+        java2Ws("faulty-hello", "FaultyHelloServiceImpl");
     }
 
     public void java2Ws(String path, String serviceName)

